File: packages/fortress/android/src/main/java/io/capkit/fortress/config/RuntimeConfigStore.kt
package io.capkit.fortress.config

import android.content.Context
import com.getcapacitor.JSObject
import org.json.JSONObject

class RuntimeConfigStore(
  context: Context,
) {
  private companion object {
    const val PREFS_NAME = "fortress_runtime_config"
    const val KEY_PAYLOAD = "payload"
    const val PAYLOAD_VERSION = 1

    val ALLOWED_KEYS =
      setOf(
        "verboseLogging",
        "logLevel",
        "lockAfterMs",
        "enablePrivacyScreen",
        "privacyOverlayText",
        "privacyOverlayImageName",
        "privacyOverlayShowText",
        "privacyOverlayShowImage",
        "privacyOverlayTextColor",
        "privacyOverlayBackgroundOpacity",
        "privacyOverlayTheme",
        "fallbackStrategy",
        "allowCachedAuthentication",
        "cachedAuthenticationTimeoutMs",
        "maxBiometricAttempts",
        "lockoutDurationMs",
        "requireFreshAuthenticationMs",
        "encryptionAlgorithm",
        "persistSessionState",
      )
  }

  private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

  fun loadOverrides(): JSObject? {
    val rawPayload = prefs.getString(KEY_PAYLOAD, null) ?: return null

    val payload = runCatching { JSONObject(rawPayload) }.getOrNull() ?: return null
    if (payload.optInt("version", -1) != PAYLOAD_VERSION) {
      return null
    }

    val rawOverrides = payload.optJSONObject("overrides") ?: return null
    return sanitize(rawOverrides)
  }

  fun saveOverrides(rawOverrides: JSObject) {
    val sanitized = sanitize(rawOverrides)
    val payload =
      JSONObject()
        .put("version", PAYLOAD_VERSION)
        .put("updatedAt", System.currentTimeMillis())
        .put("overrides", sanitized)

    prefs.edit().putString(KEY_PAYLOAD, payload.toString()).apply()
  }

  fun clearOverrides() {
    prefs.edit().remove(KEY_PAYLOAD).apply()
  }

  private fun sanitize(source: JSONObject): JSObject {
    val result = JSObject()

    val keys = source.keys()
    while (keys.hasNext()) {
      val key = keys.next()
      if (!ALLOWED_KEYS.contains(key)) {
        continue
      }

      val value = source.opt(key)
      if (value is Boolean || value is Int || value is Long || value is Double || value is String) {
        result.put(key, value)
      }
    }

    return result
  }
}

File: packages/fortress/ios/Sources/FortressPlugin/Config/RuntimeConfigStore.swift
import Foundation

struct RuntimeConfigStore {

    private static let payloadVersion = 1
    private static let storageKey = "fortress_runtime_config_v1"

    private static let allowedOverrideKeys: Set<String> = [
        "verboseLogging",
        "logLevel",
        "lockAfterMs",
        "enablePrivacyScreen",
        "privacyOverlayText",
        "privacyOverlayImageName",
        "privacyOverlayShowText",
        "privacyOverlayShowImage",
        "privacyOverlayTextColor",
        "privacyOverlayBackgroundOpacity",
        "privacyOverlayTheme",
        "fallbackStrategy",
        "allowCachedAuthentication",
        "cachedAuthenticationTimeoutMs",
        "maxBiometricAttempts",
        "lockoutDurationMs",
        "requireFreshAuthenticationMs",
        "encryptionAlgorithm",
        "persistSessionState"
    ]

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func loadOverrides() -> [String: Any]? {
        guard let payload = defaults.dictionary(forKey: Self.storageKey) else {
            return nil
        }

        guard let version = payload["version"] as? Int, version == Self.payloadVersion else {
            return nil
        }

        guard let rawOverrides = payload["overrides"] as? [String: Any] else {
            return nil
        }

        return sanitize(rawOverrides)
    }

    func saveOverrides(_ overrides: [String: Any]) {
        let sanitized = sanitize(overrides)
        let payload: [String: Any] = [
            "version": Self.payloadVersion,
            "updatedAt": Int64(Date().timeIntervalSince1970 * 1000),
            "overrides": sanitized
        ]
        defaults.set(payload, forKey: Self.storageKey)
    }

    func clearOverrides() {
        defaults.removeObject(forKey: Self.storageKey)
    }

    private func sanitize(_ source: [String: Any]) -> [String: Any] {
        var result: [String: Any] = [:]

        for (key, value) in source {
            guard Self.allowedOverrideKeys.contains(key) else { continue }

            if value is Bool || value is Int || value is Double || value is String {
                result[key] = value
            }
        }

        return result
    }
}
